### ISSUE TITLE
Request for guidance on v2->v3 migration for get_writer

### ISSUE THREAD

**Issue by hmaarrfk** [imageio/imageio#878]
So I understand the motivation behind deprecating v2 to v3.

I would like to ask for a little more guidance in going from `imageio.v2.get_writer` to new desirable API.

Maybe this API has not been formally deprecated and we should still use `get_writer` from the top level?

We use the `get_writer` API with generated data so we can't always fit it in memory.

For example an adapted snipped of code of ours might look like:

```python
import imageio
writer = imageio.v2.get_writer('test_v2.mp4', format='FFMPEG', mode='I',
                            quality=None, codec='libopenh264',
                            fps=30, output_params=['-filter:v', 'transpose=dir=cclock'])
for i in range(16):
    # Generate interesting data
    frame = np.zeros((512, 512, 3), dtype='uint8')
    writer.append_data(frame)
writer.close()
```

My attempt with v3 is the following:
```python
import imageio
writer = imageio.v3.imopen('test_v3.mp4', plugin='FFMPEG', io_mode='wI',
                            quality=None, codec='libopenh264',
                            fps=30, output_params=['-filter:v', 'transpose=dir=cclock'])
for i in range(16):
    # Generate interesting data
    frame = np.zeros((512, 512, 3), dtype='uint8')
    writer.append_data(frame)
writer.close()
```

However, this yields:
```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/mambaforge/envs/dev/lib/python3.9/site-packages/imageio/core/imopen.py in imopen(uri, io_mode, plugin, extension, format_hint, legacy_mode, **kwargs)
    146         try:
--> 147             return loader(request, **kwargs)
    148         except InitializationError as class_specific:

~/mambaforge/envs/dev/lib/python3.9/site-packages/imageio/core/imopen.py in loader(request, **kwargs)
    134             def loader(request, **kwargs):
--> 135                 return config.plugin_class(request, **kwargs)
    136

TypeError: partial_legacy_plugin() got an unexpected keyword argument 'quality'

The above exception was the direct cause of the following exception:

OSError                                   Traceback (most recent call last)
<ipython-input-1-5162b99564c1> in <cell line: 2>()
      1 import imageio
----> 2 writer = imageio.v3.imopen('test_v3.mp4', plugin='FFMPEG', io_mode='wI',
      3                             quality=None, codec='libopenh264',
      4                             fps=30, output_params=['-filter:v', 'transpose=dir=cclock'])
      5 for i in range(16):

~/mambaforge/envs/mcam_dev/lib/python3.9/site-packages/imageio/core/imopen.py in imopen(uri, io_mode, plugin, extension, format_hint, legacy_mode, **kwargs)
    163
    164         request.finish()
--> 165         raise err_type(err_msg) from err_from
    166
    167     # fast-path based on format_hint

OSError: An unknown error occured while initializing plugin `FFMPEG`.
```

```
imageio.__version__
'2.21.2'
```

Whatever the answer, it may be good to add this snippet above to the test suite ;)

**Comment by hmaarrfk** [comment]
This is mostly stopping us from doing

```diff
- import imageio
+ import imageio.v3 as imageio
```

**Comment by FirefoxMetzger** [comment]
> For example an adapted snipped of code of ours might look like:

Let me answer this one first. Here is how it works in v3 (in principle):

```python
import imageio.v3 as iio

with iio.imopen('test_v3.mp4', "w", plugin='FFMPEG') as file:
    for i in range(16):
        frame = np.zeros((512, 512, 3), dtype='uint8')
        file.write(
            frame,
            quality=None,
            codec='libopenh264',
            fps=30,
            output_params=['-filter:v', 'transpose=dir=cclock']
        )
```

That being said, there is currently a bug that prevents incremental writing (appending) in v3 via FFMPEG and instead it overwrites the file each time :(. It's conceptually easy to fix, so hopefully there will be someone interested in making a contribution soon(TM).

As an alternative, you could use the pyav/av backend to do incremental writing (which I know you know how to do, so I will omit an example for brevity :D).

> I would like to ask for a little more guidance in going from imageio.v2.get_writer to new desirable API.

Yes, the docs are a bit vague here, because the exact syntax is use-case dependant. 

On a high level, `imageio.v2.get_writer` selects a plugin (formerly format), initializes it, and returns it to you. From there, you call `plugin.append_data`  and other methods of the [old plugin API](https://imageio.readthedocs.io/en/stable/development/plugins.html#example-template-plugin). 

In v3, you call `imageio.v3.imopen`, which selects a plugin, initializes it, and returns it to you. From there you call `plugin.write` and other methods of the respective plugin class. What is available may be plugin dependent, but at a minimum you are guaranteed to find those defined in the [new plugin API](https://imageio.readthedocs.io/en/stable/_autosummary/imageio.core.v3_plugin_api.PluginV3.html#imageio.core.v3_plugin_api.PluginV3).

(Notice that one of the features currently missing is "how do I write metadata to a file". While discussing the API with those interested in discussing nobody brought up this use-case so we missed it; however, I would like to find an elegant way to add it ... suggestions welcome :D)

> Maybe this API has not been formally deprecated and we should still use get_writer from the top level?

The API in `imageio.v2` will continue to exist in v3 and there won't be a deprecation of it. That said, it will become a thin wrapper around calls to `plugin = imopen(...); plugin.XXX` that accomplish the requested task.

The two functions that are not easy to provide this way are `get_reader` and `get_writer`, because they expose internals of the old format manager that we are deprecating (a bit of bad design on our part 😅). I have an item on my todo list to check if we can write a wrapper that enables using v3 plugins with `get_reader` / `get_writer`. That said, my todo list is quite long at the moment and I have no personal use for this feature, so I didn't actually implement anything yet 😢.

The bottom line is, if we can support those functions by wrapping calls to the v3 plugin API then they will continue to exist as, e.g., `imageio.v2.get_writer`. If not, then I think we will have to depreciate them, because they are messy in terms of file-handling (ensuring we don't leak) and maintenance. That being said, `imageio.v3` should be able to accommodate any relevant use-case :)

> We use the get_writer API with generated data so we can't always fit it in memory.

For video, like in your case, I think the relevant discussion is happening here: https://github.com/imageio/imageio/pull/853#issuecomment-1245621528 (the extension to the pyav API).

**Comment by FirefoxMetzger** [comment]
@hmaarrfk Is this issue still relevant for you? If so, what is the current blocker preventing you from igrating to v3?

**Comment by hmaarrfk** [comment]
it is no longer relevant to me, mostly because I have a need to better control encoding options and had to use PyAv more directly.

**Comment by hmaarrfk** [comment]
Thank yo ufor following up!
